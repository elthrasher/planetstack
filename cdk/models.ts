import { CfnModel, WebSocketApi } from '@aws-cdk/aws-apigatewayv2';
import { Stack } from '@aws-cdk/core';

import { MessageAction } from '../types/MessageAction';

// Create `CfnModel` from properties.
const getModel = (
  scope: Stack,
  modelName: string,
  properties: Record<string, { enum?: string[]; $ref?: string; type?: string }>,
  required: string[],
  webSocketApi: WebSocketApi,
) => {
  return new CfnModel(scope, `${modelName}Model`, {
    apiId: webSocketApi.apiId,
    contentType: 'application/json',
    name: `${modelName}Model`,
    schema: {
      $schema: 'http://json-schema.org/draft-04/schema#',
      properties,
      required,
      title: `${modelName}Schema`,
      type: 'object',
    },
  });
};

// Create models for validation. It's possible to generate models from TypeScript interfaces or types.
// See https://matt.martz.codes/how-to-automatically-generate-request-models-from-typescript-interfaces for some inspiration.
export const getModels = (scope: Stack, webSocketApi: WebSocketApi): { [key: string]: CfnModel } => {
  const iconModel = getModel(
    scope,
    'Icon',
    {
      id: { type: 'string' },
      img: { type: 'number' },
      x: { type: 'number' },
      y: { type: 'number' },
    },
    ['img', 'x', 'y'],
    webSocketApi,
  );

  const addModel = getModel(
    scope,
    'Add',
    {
      action: { enum: [MessageAction.ADD_ICON] },
      icon: { $ref: `https://apigateway.amazonaws.com/restapis/${webSocketApi.apiId}/models/${iconModel.ref}` },
    },
    ['action', 'icon'],
    webSocketApi,
  );

  const changeBgModel = getModel(
    scope,
    'ChangeBackground',
    { action: { enum: [MessageAction.CHANGE_BACKGROUND] }, bg: { type: 'number' } },
    ['action', 'bg'],
    webSocketApi,
  );

  const deleteModel = getModel(
    scope,
    'Delete',
    { action: { enum: [MessageAction.DELETE_ICON] }, id: { type: 'string' } },
    ['action', 'id'],
    webSocketApi,
  );

  const getStateModel = getModel(
    scope,
    'GetState',
    { action: { enum: [MessageAction.GET_STATE] } },
    ['action'],
    webSocketApi,
  );

  const moveModel = getModel(
    scope,
    'Move',
    {
      action: { enum: [MessageAction.MOVE_ICON] },
      icon: { $ref: `https://apigateway.amazonaws.com/restapis/${webSocketApi.apiId}/models/${iconModel.ref}` },
    },
    ['action', 'icon'],
    webSocketApi,
  );

  return { addModel, changeBgModel, deleteModel, getStateModel, moveModel };
};
