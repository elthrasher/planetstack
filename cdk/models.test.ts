import { WebSocketApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { getModels } from './models';

describe('models', () => {
  test('models', () => {
    const app = new App();
    const stack = new Stack(app);
    getModels(stack, new WebSocketApi(stack, 'testApi'));
    const cfn = Template.fromStack(stack).toJSON();

    expect(cfn).toMatchSnapshot();
  });
});
