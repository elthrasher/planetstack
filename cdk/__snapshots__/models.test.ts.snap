// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`models models 1`] = `
Object {
  "Resources": Object {
    "AddModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "AddModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "addIcon",
              ],
            },
            "icon": Object {
              "$ref": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "https://apigateway.amazonaws.com/restapis/",
                    Object {
                      "Ref": "testApiD6ECAB50",
                    },
                    "/models/",
                    Object {
                      "Ref": "IconModel",
                    },
                  ],
                ],
              },
            },
          },
          "required": Array [
            "action",
            "icon",
          ],
          "title": "AddSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "ChangeBackgroundModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "ChangeBackgroundModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "changeBackground",
              ],
            },
            "bg": Object {
              "type": "number",
            },
          },
          "required": Array [
            "action",
            "bg",
          ],
          "title": "ChangeBackgroundSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "DeleteModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "DeleteModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "deleteIcon",
              ],
            },
            "id": Object {
              "type": "string",
            },
          },
          "required": Array [
            "action",
            "id",
          ],
          "title": "DeleteSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "GetStateModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "GetStateModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "getState",
              ],
            },
          },
          "required": Array [
            "action",
          ],
          "title": "GetStateSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "IconModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "IconModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "id": Object {
              "type": "string",
            },
            "img": Object {
              "type": "number",
            },
            "x": Object {
              "type": "number",
            },
            "y": Object {
              "type": "number",
            },
          },
          "required": Array [
            "img",
            "x",
            "y",
          ],
          "title": "IconSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "MoveModel": Object {
      "Properties": Object {
        "ApiId": Object {
          "Ref": "testApiD6ECAB50",
        },
        "ContentType": "application/json",
        "Name": "MoveModel",
        "Schema": Object {
          "$schema": "http://json-schema.org/draft-04/schema#",
          "properties": Object {
            "action": Object {
              "enum": Array [
                "moveIcon",
              ],
            },
            "icon": Object {
              "$ref": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "https://apigateway.amazonaws.com/restapis/",
                    Object {
                      "Ref": "testApiD6ECAB50",
                    },
                    "/models/",
                    Object {
                      "Ref": "IconModel",
                    },
                  ],
                ],
              },
            },
          },
          "required": Array [
            "action",
            "icon",
          ],
          "title": "MoveSchema",
          "type": "object",
        },
      },
      "Type": "AWS::ApiGatewayV2::Model",
    },
    "testApiD6ECAB50": Object {
      "Properties": Object {
        "Name": "testApi",
        "ProtocolType": "WEBSOCKET",
        "RouteSelectionExpression": "$request.body.action",
      },
      "Type": "AWS::ApiGatewayV2::Api",
    },
  },
}
`;
